### PR TITLE
ID-1230 MAU, MEU Event Logging

### PR DESCRIPTION
**Pull request by tlangs**
Ticket: https://broadworkbench.atlassian.net/browse/ID-1230

Replaces Sam's "OIDCHeader" log entries with entries that have a queryable JSON Payload.

Because this replaces an existing log message, there should be no addition to log volume. 

```
jsonPayload: {
  @timestamp: "2024-05-03T19:07:59.395Z",
  @version: "1",
  eventProperties: {
    warehouse: true,
    event: "apiRequest:user:complete",
    request: {
      method: "GET",
      oidcHeaders: {
        azureB2CId: "41eab911-788f-4f99-bed1-098609746ade",
        email: "tlangs.dev@gmail.com",
        googleSubjectId: "112763693687848748453",
        managedIdentityObjectId: null,
      },
      path: "/register/user/v2/self/info"
    },
  response: {
    status: 200
  },
  samUserId: "271467912109162ba84c7"
}
```

---

**PR checklist**

- [x] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've filled out the [Security Risk Assessment](https://sdarq.dsp-appsec.broadinstitute.org/jira-ticket-risk-assesment) (requires Broad Internal network access) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala [hunk 1]
```diff
@@ -29,33 +40,42 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
   def withActiveUser(samRequestContext: SamRequestContext): Directive1[SamUser] = requireOidcHeaders.flatMap { oidcHeaders =>
     onSuccess {
       getActiveSamUser(oidcHeaders, userService, termsOfServiceConfig, samRequestContext).unsafeToFuture()
-    }.tmap { samUser =>
+    }.flatMap { samUser =>
       logger.debug(s"Handling request for active Sam User: $samUser")
-      samUser
+      logRequestResultWithSamUser(samUser, oidcHeaders)
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala:45]: By sticking the logging directive here, we can log all api requests with an active user.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala [hunk 2]
```diff
@@ -29,33 +40,42 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
...
+      .mapInnerRoute { r =>
+        if (!adminConfig.serviceAccountAdmins.contains(oidcHeaders.email)) {
+          reject(AuthorizationFailedRejection)
+        } else {
+          logger.info(s"Handling request for service admin account: ${oidcHeaders.email}")
+          r
+        }
       }
-    }
+      .tflatMap(_ => logAdminServiceAdminUserRequestResult(oidcHeaders))
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala:59]: Same here, but for service admins like thurloe

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala [hunk 3]
```diff
@@ -29,33 +40,42 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
...
   }
 
   def withUserAllowInactive(samRequestContext: SamRequestContext): Directive1[SamUser] = requireOidcHeaders.flatMap { oidcHeaders =>
     onSuccess {
       getSamUser(oidcHeaders, userService, samRequestContext).unsafeToFuture()
-    }.tmap { samUser =>
+    }.flatMap { samUser =>
       logger.debug(s"Handling request for (in)active Sam User: $samUser")
-      samUser
+      logRequestResultWithSamUser(samUser, oidcHeaders)
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala]: This one catches requests by inactive users.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala [hunk 4]
```diff
@@ -75,7 +95,7 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
       optionalHeaderValueByName(managedIdentityObjectIdHeader).map(_.map(ManagedIdentityObjectId)))
       .as(OIDCHeaders)
       .map { oidcHeaders =>
-        logger.info(s"Auth Headers: $oidcHeaders")
+        logger.debug(s"Auth Headers: $oidcHeaders")
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala:102]: This makes Sam no longer log the "Auth Headers" message when deployed. Its replaced by the new log message with json payload.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala [hunk 5]
```diff
@@ -85,6 +105,63 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
       _ => Left(GoogleSubjectId(idString)) // id is a number which is what google subject ids look like
     )
   }
+
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala:112]: Here's the meat of the changes. These are what get called on every requests

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala (hunk 3 above)
Comment (Ghost-in-a-Jar) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala]: Is there a way to differentiate between active/inactive in the logs? Maybe we just want to leave this one out for now if they wont count for MAU/MEU?

**Review by Ghost-in-a-Jar — APPROVED** [review]

**Review by samanehsan — APPROVED** [review]

**Pull request merged**